### DISCUSSION TITLE
Can mounting an untrusted root be avoided?

### DISCUSSION THREAD

**Discussion by Ablu** [uapi-group/specifications#87] [Q&A]
I am wondering: Assuming I want a dm-verity `/usr` and a read-writable, encrypted `/` (auto-created by systemd-repart). Then, when transitioning from an UKI to the final rootfs. How can it be ensured that this rootfs is genuine?

I understand that encryption keys of the rootfs can be measured into TPM, disallowing access to secrets if that filesystem is not there. But how can auto-discovering and mounting a maliciously forged partition be avoided? Is there a mechanism that allows the initramfs to reject such a partition?

**Comment by bluca** [comment]
Why would it matter?

**Reply by daandemeyer** [reply]
Because I don't want my system running untrusted software without my permission? Let's say this untrusted root partition starts doing malicious activities like joining a botnet or something like that, that seems like something we'd want to avoid no?

**Reply by bluca** [reply]
What difference does it make if it's the root partition or the esp that is substituted? What's the threat model?

**Reply by daandemeyer** [reply]
Substituting the ESP should not be an issue if Secure Boot + UKI is used because we still won't boot into an untrusted UKI, and all the rest is authenticated before use as well. 

The threat model is that someone adds another root partition to my unattended server, triggers a reboot and my server suddenly starts running malicious things. I want my unattended server to be running my workload or no workload at all.

**Reply by ElvishJerricco** [reply]
@bluca As I tried to outline in https://github.com/systemd/systemd/issues/28491, common configurations, including gpt-auto, can have one disk decrypted by initrd, when another disk ends up being booted. There are things you can do to help (well, not so much with the way gpt-auto works) but it's a problem IMO

**Reply by bluca** [reply]
> Substituting the ESP should not be an issue if Secure Boot + UKI is used because we still won't boot into an untrusted UKI, and all the rest is authenticated before use as well.
> 
> The threat model is that someone adds another root partition to my unattended server, triggers a reboot and my server suddenly starts running malicious things. I want my unattended server to be running my workload or no workload at all.

So the threat model is unauthorized code running - then the solution is code integrity. Up to the kernel that is done via secure boot as you said, after the kernel there is no mainline solution. There is IPE that provides it for userspace binaries, but it is not merged yet. The only current partial solution is remote attestation, which is after-the-fact checks.
The partition is irrelevant though, unauthorized code can come from any sources, at any time.

**Reply by berrange** [reply]
> @bluca As I tried to outline in https://github.com/systemd/systemd/issues/28491, common configurations, including gpt-auto, can have one disk decrypted by initrd, when another disk ends up being booted. There are things you can do to help (well, not so much with the way gpt-auto works) but it's a problem IMO

This is not a new threat with gpt auto-discovery. Prior to auto-discovery the rootfs was typically hardcoded by specifying its UUID on the kernel command line, which was visible as cleartext. So an attacker could just give their replacement rootfs the same UUID as the original one.

**Reply by bluca** [reply]
This is IPE: https://microsoft.github.io/ipe/
Last revision from last week under review: https://lore.kernel.org/all/1696457386-3010-1-git-send-email-wufan@linux.microsoft.com/

**Reply by daandemeyer** [reply]
> The partition is irrelevant though, unauthorized code can come from any sources, at any time.

It'd still be nice to ignore any root partition without the right properties (not encrypted against signed PCRs) so that the server can keep doing its thing even if someone tries to mess with it by adding a second root partition. On the other hand, failing hard if this happens might be better, since any kind of tampering is probably a reason to fail hard.

**Reply by Ablu** [reply]
I disagree that this is only about prevent unauthenticated code running.

Let's assume the following:
`/` is a dm-verity partition. A persistent machine-id comes from _somewhere_. It looks like I can use roothash to make sure that root is not tampered with.
My system needs some state. So I have a LUKS-backend `/var`.
Now, how - in a discoverable partition scenario - can I prevent an attacker to replace that `/var` with different runtime-data?

**Reply by bluca** [reply]
The challenge is having a way to "indicate" that, that also can't be faked in the same threat model, which is physical and full write access. And also only partially solves the problem, again unauthorized code can come from many places, and physical write access is probably one of the hardest vectors already anyway

**Reply by poettering** [reply]
> The threat model is that someone adds another root partition to my unattended server, triggers a reboot and my server suddenly starts running malicious things. I want my unattended server to be running my workload or no workload at all.

the image policy stuff (`systemd.image_policy=root=encrypted` on the kernel cmdline) is supposed to allow you to say "boot an encrypted disk or not at all". gpt-auto-generator takes that into consideration. Thus, "someone" can't really add such a root partition that the system would just boot up, since that someone would have to bind it to the TPM too, or know your passphrase/recovery key, so that they can trick you into unlocking it at boot.

**Reply by Ablu** [reply]
> the image policy stuff (systemd.image_policy=root=encrypted on the kernel cmdline) is supposed to allow you to say "boot an encrypted disk or not at all". gpt-auto-generator takes that into consideration. Thus, "someone" can't really add such a root partition that the system would just boot up, since that someone would have to bind it to the TPM too, or know your passphrase/recovery key, so that they can trick you into unlocking it at boot.

THAT sounds exactly like the puzzle piece that I was missing! Thanks a lot!

**Reply by bluca** [reply]
They can though - it can be swapped with a different partition, still encrypted, say for example against PCR7 or some other known and predictable quantities. It would still happily run. So again, it goes back to remote attestation to check which volume key is used, but again that's after the fact.

**Reply by daandemeyer** [reply]
Shouldn't we extend the image policy to allow encoding that the root partition should be encrypted against an expected set of PCRs? So that we enforce that it's encrypted against the local stuff as well to make it as hard as possible to craft a root partition that we'd use

**Reply by ElvishJerricco** [reply]
@DaanDeMeyer I don't think it's possible to do that to a helpful degree? Everything before the decryption of the root partition is, obviously, unencrypted. So the PCRs are fundamentally *all* predictable (even if the firmware measurements are a little tricky to get at)

**Reply by bluca** [reply]
A lot of the local stuff, if you mean machine id, volume key etc, come after the root partition, so chicken and egg

**Reply by Ablu** [reply]
> They can though - it can be swapped with a different partition, still encrypted, say for example against PCR7 or some other known and predictable quantities. It would still happily run. So again, it goes back to remote attestation to check which volume key is used, but again that's after the fact.

Hm. I fear I do not know all the concepts well enough to understand this. How do I forge a LUKS partition that will be accepted in place of my original one?

**Reply by bluca** [reply]
> > They can though - it can be swapped with a different partition, still encrypted, say for example against PCR7 or some other known and predictable quantities. It would still happily run. So again, it goes back to remote attestation to check which volume key is used, but again that's after the fact.
> 
> Hm. I fear I do not know all the concepts well enough to understand this. How do I forge a LUKS partition that will be accepted in place of my original one?

The policy simply says "this must be encrypted". We are talking about unattended cases like servers, so realistically that will be against some PCRs - but those are very easy to predict, so right now you just encrypted against the expected value of PCR7 and everything will happily load, even with that policy. Even with the enhancements proposed here, to have a policy that mentions a PCR, that still doesn't help much for the root case, as the threat model assumes full physical access, so one can check what the measurements of all PCRs are, and encrypt against that.

**Reply by Ablu** [reply]
> The policy simply says "this must be encrypted". We are talking about unattended cases like servers, so realistically that will be against some PCRs - but those are very easy to predict, so right now you just encrypted against the expected value of PCR7 and everything will happily load, even with that policy. Even with the enhancements proposed here, to have a policy that mentions a PCR, that still doesn't help much for the root case, as the threat model assumes full physical access, so one can check what the measurements of all PCRs are, and encrypt against that.

Thanks for the explanation! Since I probably do not know enough about TPMs and LUKS I just want to confirm:

The idea is that since there is no authenticity data of the partition is required for unlocking key, the attacker can just predict the TPM state and derive a key that will be accepted?

Is there no secret from the TPM required for that? Or does it require some degree of code execution for creating the key for the fake partition?

**Reply by bluca** [reply]
Yes it requires access to the TPM, but in this threat model there is full physical access so that can be obtained, otherwise changing the disk is also impossible

**Reply by Ablu** [reply]
Sure! Just wanted to make sure I understand it correctly. Thanks a lot!

**Reply by Ablu** [reply]
I have unmarked this as "solved" then for now.

**Comment by poettering** [comment]
I am not sure I understand the question. What does "genuine" mean? If the OS booted up, found no existing root fs, and created one, locking it to the TPM (which means encrypting the LUKS unlock key with a key that is only known to the TPM), then you will need the TPM to get access. If some other root fs comes along then it will not have its unlock key encrypted with that TPM's encryption key, and thus cannot be unlocked. Except of course if you have *two* images that both were created initially on the local system. But in that case both are equally "genuine".

It might make sense to authenticate the OS to the user logging in (traditionally it's the user who authenticates to the OS, but not vice versa). This is a form of "remote attestation" and is typically done via TOTP: for example with stuff like this:

https://github.com/mjg59/tpmtotp

The idea is basically that at login time you get shown a code that proves for the current time that the OS in order and is the system you think it is. You can verify that with your phone.

We have discussed a couple of times, whether we should add something like this to systemd itself, to just be there out of the box, but I am not sure if this isn't too far in "nerd shit" territory.

**Reply by Ablu** [reply]
> If some other root fs comes along then it will not have its unlock key encrypted with that TPM's encryption key, and thus cannot be unlocked.

My core question is how this "cannot be unlocked" is done. From my understanding of the [discoverable partition](https://uapi-group.org/specifications/specs/discoverable_partitions_specification/) part, wouldn't a non-encrypted partition also be discovered and used? Is there some mechanism that allows me to prevent that?

**Reply by poettering** [reply]
https://www.freedesktop.org/software/systemd/man/systemd.image-policy.html

**Comment by poettering** [comment]
> I understand that encryption keys of the rootfs can be measured into TPM, disallowing access to secrets if that filesystem is not there.

So we do measure the LUKS volume key to some TPM PCR. But generally, this PCR is not used for unlocking the disk, because that would be a cyclic dependency then: you'd have to have measured to the volume key to acquire the volume key.

**Comment by poettering** [comment]
note that the image policy stuff we probably should extend a bit, and allow encoding some rules on *how* an encrypted disk is unlocked, so that you can declare that unlocking via pw is never allowed.

i.e. `systemd.image_policy=root=encrypted-tpm2+encrypted-fido2` would dictate that you can only unlock via tpm2 or fido2, but not via pw or so. Happy to take patches for that ;-)

**Reply by bluca** [reply]
That would certainly be useful, but still wouldn't help with this specific use case - assuming unrestricted physical write access, creating a 'malicious' partition encrypted against say PCR7 or so is just as trivial, and then it would pass that policy. But again, the key thing is defining the threat model and what it needs to protect against. Normally the concern with physical attacks in datacenters is more around data exfiltration. If the concern is code integrity, then the partition is just one of many vectors, and not even remotely the easiest one to exploit.

**Reply by poettering** [reply]
> creating a 'malicious' partition encrypted against say PCR7 or so is just as trivial

is it? you'd need to have access to the TPM for that.

**Reply by bluca** [reply]
The threat model defined here assumes full physical access (eg: to swap disks), so you run whatever you want to get what you need, and then prepare the vector

**Reply by ElvishJerricco** [reply]
which of course requires either a stock OS signed by a vendor, or a firmware that allows someone with physical access to freely disable secure boot

**Reply by bluca** [reply]
Not a whole OS, just a kernel from any of the vendors that get signed by the UEFI 3rd party CA, of which there is many

**Comment by Vogtinator** [comment]
IMO this is a valid concern and I also raised that with the the related GRUB TPM protector development, resulting in (so far) https://lists.gnu.org/archive/html/grub-devel/2023-10/msg00025.html

The best way to protect against this would be to use authenticated encryption with the authentication key measured before attempting to unlock.

**Reply by poettering** [reply]
> The best way to protect against this would be to use to use authenticated encryption with the authentication key measured before attempting to unlock.

Note that systemd-cryptsetup can measure the volume key for you on unlock already, see the `tpm2-measure-pcr=` crypttab option. This is in fact enabled by default if you use the systemd-gpt-auto-generator logic on a UKI boot. 

systemd-repart currently doesn't support formatting a disk with authenticated encryption (i.e. dm-crypt + dm-integrity), but we certainly should add that.

**Reply by Vogtinator** [reply]
> systemd-repart currently doesn't support formatting a disk with authenticated encryption (i.e. dm-crypt + dm-integrity), but we certainly should add that.

Yeah. FWICT this is handled rather transparently by cryptsetup, so the existence of dm-integrity is mostly an implementation detail.

**Reply by Ablu** [reply]
Do I understand this correctly, that if systemd-repart could set up the authenticated encryption then my described scenario would be solved? Since revealing the unlock key would be sealed against the authentication key I would no longer be able to forge a different partition by predicting TPM state (as sketched in https://github.com/orgs/uapi-group/discussions/87#discussioncomment-7271771)?

**Comment by poettering** [comment]
@bluca btw, there's one more thing we need: with Dan's support for encrypting disks against an SRK offline we can have really nice confext DDIs that can be prepared offline and then can only be unlocked on a specific system. But for this to be useful we'd also need the ability to *sign* the image with some admin key, so that the system can authenticate things before it applies the configuration.
